part4/advanced-customizations.md

```diff
@@ -18,4 +18,63 @@ TODO
 Customizing Message Handler behavior
 ------------------------------------
 
-TODO
+HanderEnhancers
+---------------
+
+It is possible to wrap handlers with custom logic. This differs from HandlerInterceptors in that you have access to the Aggregate member at the time of resolving.
+You can use HandlerEnhancers to intercept and perform checks on groups of `@CommandHandler`s or `@EventHandler`s. 
+
+To create a HandlerEnhancer you start by implementing `HandlerEnhancerDefinition` and overriding the `wrapHandler()` method. 
+All this method does is give you access to the `MessageHandlingMember<T>` which is an object representing any handler that is specified in the system. 
+
+You can then sort these handlers based on their annotations by using the `annotationAttributes(Annotation annotation)` method. This will filter out only those handlers that use that `Annotation`.
+
+For your HandlerEnhancer to run you'll need to create a `META-INF/services/org.axonframework.messaging.annotation.HandlerEnhancerDefinition` file containing
+the fully qualified name of the handler enhancer you would like to use.
+
+Example:
+
+```java
+public class MethodCommandHandlerDefinition implements HandlerEnhancerDefinition { // 1.
+
+    @Override
+    public <T> MessageHandlingMember<T> wrapHandler(MessageHandlingMember<T> original) { // 2.
+        return original.annotationAttributes(CommandHandler.class) // 3.
+                .map(attr -> (MessageHandlingMember<T>) new MethodCommandMessageHandlingMember(original, attr))
+                .orElse(original); // 5.
+    }
+
+    private static class MethodCommandMessageHandlingMember<T> extends WrappedMessageHandlingMember<T>{
+
+        private final String commandName;
+
+        private MethodCommandMessageHandlingMember(MessageHandlingMember<T> delegate,
+                                                   Map<String, Object> annotationAttributes) {
+            super(delegate);
+
+            if ("".equals(annotationAttributes.get("commandName"))) {
+                commandName = delegate.payloadType().getName();
+            } else {
+                commandName = (String) annotationAttributes.get("commandName");
+            }
+        }
+
+        @Override
+        public boolean canHandle(Message<?> message) {
+            return super.canHandle(message) && commandName.equals(((CommandMessage) message).getCommandName()); // 4.
+        }
+
+        public String commandName() {
```

Comment:
Shouldn't this getter be called `getCommandName()`? At least that's how it's being called.